---
title: "Simulation via AlphaSimR"
author: "MaoHuang"
date: "4/27/2022"
output:
  html_document: default
  pdf_document: default
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>


## Objectives
Class1: <br>
We will learn different `AlphaSimR` functions <br>
We will write chunks of codes to simulate different scenarios

Class2: <br>
We will write whole breeding pipeline <br>
We will evaluate different results as we change the breeding scheme

## Additional Reading materials
Reading materials for <br>
*  Writing scripts via [R Markdown](https://rmarkdown.rstudio.com/authoring_quick_tour.html)  <br>
*  Create project in [Rstudio](https://support.rstudio.com/hc/en-us/articles/200526207-Using-RStudio-Projects)  <br>
*  [Github](https://r-pkgs.org/git.html#git-rstudio) commit and push  <br>

## General steps to use `AlphaSimR`
1. Plan out your breeding program  <br>
2. Set up founder haplotypes   <br>
3. Set up simulation parameters  <br>
4. Modeling the breeding program  <br>
5. Evalute your simulation results  <br>

## Script setup  
#### Install packages  
*** Note: `AlphaSimR` generates many random numbers. Set up random seed: your will always get the same results. Please change it when you run.

```{r Script setup}  
## echo=FALSE in R markdown, if you do not want to show this chunk of code
random_seed <- 12345
set.seed(random_seed)

## Install the package if haven't yet
packages_used <- c("AlphaSimR")
packages_installed <- installed.packages()  # what's been installed
for (package in packages_used){
  if (!(package %in% packages_installed[,"Package"])){
    install.packages(package)
  } else {
  library(package,character.only=T)
       }
}
```

#### Set up `AlphaSimR` parameters  
Simulation uses assumptions <br>
*Mendel’s law of segregation <br>
*Mendel’s law of independent assortment <br>
*GxE exists <br>

AlphaSimR requires initial simulation parameters to be set up for haplotypes
```{r script parameters}
nFounders <- 10
nChr <- 20
segSites <- 500   # Number of segrerating sites per chormosome 
nQTL<- 100   # Number of QTL per chormosome 
```
The initial number of founders `nFounders` =`r nFounders` <br>
The number of chromosomes for the species is `nChr` = `r nChr`  <br>
The number of segregating sites on each chromosome is `segSites` = `r segSites`   <br>
The number of QTLs (causal variant) on each chromosome is `nQTL` = `r nQTL`  <br>


#### Some common functions in `AlphaSimR`
##### How to create founders
_Note: Example modified from [AlphaSimR](https://cran.r-project.org/web/packages/AlphaSimR/AlphaSimR.pdf), Chris Gynor [example](https://cran.r-project.org/web/packages/AlphaSimR/vignettes/intro.html) 
and [Huang](https://academic.oup.com/g3journal/article/12/3/jkac003/6511442?login=true) et al. 2022_  <br> 

* 1st, you produce founder haplotypes from a coalescent simulation as part of `AlphaSimR`. We use 
`runMacs2()` (Markovian Coalescent Simulator, [Chen](https://genome.cshlp.org/content/19/1/136) et al. 2008). It simulates bi-allelic genome sequences according to a population demographic history. <br>

You can specify predefined population histories with parameters you choose
Or supply their own population history. A list of available population histories include: GENERIC, CATTLE, WHEAT, MAIZE and EUROPEAN. <br>
You can also import your own haplotypes being generated in another software or taken directly from real marker data using the newMapPop function. <br>
Note: you can also use `quickHaplo()` function to generate the initial haplotypes by randomly sampling 1s and 0s. It is modeling a population in Hardy-Weinberg equilibrium with allele frequencies of 0.5. It is only recommended for prototype coding.  

* 2nd, you define simulation parameters that connects the founder haplotype genotypic and phenotypic variation <br>

* 3rd, your make the diploid founders from their haplotypes <br>

##### Example: <br>
Create founder haplotypes using runMacs2(); Default effective population size Ne is 100 
```{r make founder haplotypes}
founderHap<-runMacs2(nInd=nFounders,nChr=nChr,segSites=segSites)
```

Create New global simulation parameters from founder haplotypes using `SimParam$new()' 
```{r global simulation parameters}
SP<-SimParam$new(founderHap)
```
_Careful with your naming, avoid using "SP" in any other places: `AlphaSimR` other functions searches your R global environment for the object named "SP" by default ._ <br>

Add additive trait architecture. This trait is conroled by nQTL=100 QTLs per chromosome. Genetic mean is 0 and genetic variance is 1. <br>

_`addTraitA()` radomly assign eligible QTLs for additive effects ONLY. You can simulation other effects dominance (“D”), epistasis (“E”), and genotype-by-environment (“G”). You can also simulate more than one trait---If simulating more than one trait, all traits will be pleiotrophic with correlated additive effects_
```{r adding traits, etc onto global simulation parameters}
SP$addTraitA(nQtlPerChr=nQTL,mean=0,var=1) 
```

Design a SNP chip that can be used later (randomly assign eligible SNPs to a SNPchip)
```{r}
SP$addSnpChip(nSnpPerChr=200)  # Number of SNP markers per chormosome  ### Note, This could also be run at the beginning where you specify global simulation parameters

# Changes how sexes are determined in the simulation. The value "yes_sys" will systematically assign sexes to newly created individuals as first male and then female
#SP$setSexes("yes_sys")

# Track the population records 
SP$setTrackRec(TRUE)
```

```{r make founder diploids}
#3. Create a new populations of 5 individuals
founders<- newPop(founderHap, simParam=SP)
```

##### _Let's take a look at the founders_
```{r Check out the founder population}
str(founders)
```
Each individual has ID in `@id`, they are in numeric form. The `@mother` and `@father` IDs are referring to diploid parents, which are not available and are all in 0s, because this population is made from haplotypes. It gives `@gv` which is _genetic value_ or the _true breeding value_ calculated for the trait simulated from `addTraitA()`. You can access it by `gv(founders)` or `founders@gv`

##### How to access the population information
The `founders` is now your founder population object. Treat it like a special vector. Now you can extract information. For example, you can sort the individuals based on the order of their _genetic values_ and pick the top _n_ of them.
```{r extracting founders info}
# The gv is in a vector
GV<-founders@gv

# Sort individuals. Here SortID is also a population object (the sorted founders). 
Sortfounders<-founders[order(-GV)]  

# Pick top n of them
n<-3
TopID<-Sortfounders[1:n]
### str(SortID)

# OR If you want to look at all of the individuals in a dataframe format
IDGV<-as.data.frame(cbind(Sortfounders@id,Sortfounders@gv))
names(IDGV)<-c("ID","GV")
```

##### How to make some progenies
Outcrossing with `makeCross` and Selfing with `self` 

* For example, we can make some inbreds. Selfing each individual in the population with `self()`. You can control the total number of selfed progenies per parent with the parameter `nProgeny`. You can do single seed descent using this function with `nProgeny=1`. This function only works when sexes is "no" in the population.

Selfing the founder individuals for `nSelf` generations
```{r making selfing progenies}
nSelf<-6
Inbred<-founders
for (i in 1:nSelf){
  Inbred<-self(Inbred, nProgeny=2)
   #print(Inbred@id)
}
```

##### How to phenotype these individuals. 
```{r phenotyping inbreds}
### Phenotype 640 individuals of them   
varE<-1
PhenoInbred<-setPheno(Inbred,varE=varE,simParam=SP)  # h2=0.5, Becasue varG is default 1 and varE=1 
```

> * Practice: Can you select the Top10 best Inbreds based on phenotypic data?

* We can now try randomly cross n individuals selected from the inbreds (*** Note: no sexes indicated) using `makeCross`. Each being crossed once.
```{r outcrossing founders}
nSel<-10
SampleInbred<-sample(Inbred@id,nSel)   # Randomly select nSel individuals
CrossingPlan<-matrix(SampleInbred,nrow=nSel/2,ncol=2)   # A matrix with column for Female and Male, corresponding to the ids of population you are making crosses with
F1<-makeCross(Inbred,crossPlan=CrossingPlan,simParam=SP) # This is the F1 generation crossed from the two inbreds 
F2pop<-self(F1,nProgeny=100)  # Generated 100 F2s
```

##### How to build a Genomic Selection model
*  Note: Inbred pop was phenotyped above and can be genotyped. Its marker data by default is the "1st" SNP chip previously designed as part of the simulation parameters SP. If you have different SNP chips, you have to specify which SNp chip to use using intergers for the `snpChip`. So now we can train a genomic selection model using these Inbreds' phenotypic and genotypic data. We will then apply this GS model on F2s. <br> 
* Note: You can use `pullQtlGeno()` function to pull out genotypic data of your population(s).

```{r genomic selection}
GSmodel<-RRBLUP(PhenoInbred,snpChip=1,simParam=SP)  ## Set up GS model
GEBVF2<-setEBV(F2pop,GSmodel,simParam=SP) ## Estimate the GEBVs of F2s
```

##### How to compare genetic values (True BV) and breeding values
```{r}
plot(gv(GEBVF2), ebv(GEBVF2), pch=16, xlab="Genetic value", ylab=" GEBV", main="Genomic Estimated Breeding Value VS Genetic Value", cex.axis=1.3, cex.lab=1.3)
```

* We can estimate the GS accuracy by correlating the GEBV to true BV 
```{r GS accuracy}
cor(gv(GEBVF2),ebv(GEBVF2))
```

>* Practice: Can you try to set up scripts to compare GS accuracy and Phenotypic selection accuray on the inbreds? For example, we set phenotypic data for these GEBVF2s. We then select best 5% individuals based on phenotypic data vs based on their GEBVs, we then phenotype these individuals again for another year, and compare their selection accuracies. 

##### Set up a scheme to compare phenotypic accuracy and genotypic accuracy
```{r}
PhenoF2<-setPheno(GEBVF2,varE=1,simParam=SP)
nSelect<-0.1*nInd(GEBVF2)  # select 5% individuals

PSTop10<-PhenoF2[order(-PhenoF2@pheno)][1:nSelect]
GSTop10<-PhenoF2[order(-PhenoF2@ebv)][1:nSelect]

PSTop10_yr2<-setPheno(PSTop10,varE=1,simParam=SP)
GSTop10_yr2<-setPheno(GSTop10,varE=1,simParam=SP)

cor(PSTop10_yr2@pheno,PSTop10@pheno)
cor(GSTop10_yr2@pheno,GSTop10@ebv)

```

##### How to estimate selection intensity
Selection intensity is estimated as the difference of `mean(selected population)-mean(reference population)`
```{r}
## Pull out selectInd, estimate the selection intensity 

Refpop<-data.frame(PhenoF2@id,PhenoF2@pheno,PhenoF2@mother,PhenoF2@father)
colnames(Refpop)[1:2]<-c("id","trait")
Refpop$mean<-Refpop$trait/sd(Refpop$trait)

selectpop<-GSTop10@id

selectPopmean<-mean(Refpop[Refpop$id%in%selectpop,]$mean)
selectInt<-selectPopmean-mean(Refpop$mean)
print(selectInt)
```

>* Practice, Change the number of individuals being selected from 5% to 10% and report the selection intensity





